internal/tofu/context_functions.go

```diff
@@ -3,55 +3,132 @@ package tofu
+			Severity: hcl.DiagError,
+			Summary:  "Unknown function provider",
+			Detail:   fmt.Sprintf("Provider %q does not exist within the required_providers of this module", pf.Name),
+			Subject:  rng.ToHCL().Ptr(),
+		})
+	}
+
+	// Very similar to transform_provider.go
+	absPc := addrs.AbsProviderConfig{
+		Provider: pr.Type,
+		Module:   mc.Path,
+		Alias:    pf.Alias,
+	}
+
+	provider := ctx.Provider(absPc)
 
+	if provider == nil {
+		// Configured provider (NodeApplyableProvider) not required via transform_provider.go.  Instead we should use the unconfigured instance (NodeEvalableProvider) in the root.
+
+		// Make sure the alias is valid
+		validAlias := pf.Alias == ""
+		if !validAlias {
+			for _, alias := range pr.Aliases {
+				println(alias.Alias)
+				if alias.Alias == pf.Alias {
+					validAlias = true
+					break
+				}
+			}
+			if !validAlias {
+				return nil, diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unknown function provider",
+					Detail:   fmt.Sprintf("Provider %q does not have alias %q", pf.Name, pf.Alias),
+					Subject:  rng.ToHCL().Ptr(),
+				})
+			}
+		}
+
+		provider = ctx.Provider(addrs.AbsProviderConfig{Provider: pr.Type})
 		if provider == nil {
-			log.Printf("[TRACE] tofu.contextFunctions: Initializing function provider %q", addr)
-			provider, err = factory()
+			// This should not be possible
+			return nil, diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Uninitialized function provider",
+				Detail:   fmt.Sprintf("Provider %q has not yet been initialized", absPc.String()),
+				Subject:  rng.ToHCL().Ptr(),
+			})
 		}
-		return provider, err
 	}
-}
 
-// Loop through all functions specified and build a map of name -> function.
-// All functions will use the same lazily initialized provider instance.
-// This instance will run until the application is terminated.
-func providerFunctions(addr addrs.Provider, funcSpecs map[string]providers.FunctionSpec, factory providers.Factory) map[string]function.Function {
-	lazy := lazyProviderInstance(addr, factory)
-
-	functions := make(map[string]function.Function)
-	for name, spec := range funcSpecs {
-		log.Printf("[TRACE] tofu.contextFunctions: Registering function %q in provider type %q", name, addr)
-		if _, ok := functions[name]; ok {
-			panic(fmt.Sprintf("broken provider %q: multiple functions registered under name %q", addr, name))
+	// First try to look up the function from provider schema
+	schema := provider.GetProviderSchema()
+	if schema.Diagnostics.HasErrors() {
+		return nil, schema.Diagnostics
+	}
+	spec, ok := schema.Functions[pf.Function]
+	if !ok {
+		// During the validate operation, providers are not configured and therefore won't provide
+		// a comprehensive GetFunctions list
+		// Validate is built around unknown values already, we can stub in a placeholder
+		if op == walkValidate {
+			// Configured provider functions are not available during validate
```

Comment:
Do I understand correctly that this will actually only happen for **dynamically added** functions of configured providers, and will work fine for the ones that statically resolve during getschema?

Reply:
Correct!